### PR TITLE
Add Hamming Distance

### PR DESCRIPTION
**Pull request by acalejos**
Added Hamming Distance. Compared implementation to scipy implementation found at https://github.com/scipy/scipy/blob/v1.9.3/scipy/spatial/distance.py#L671-L722

The main difference is mine does allow for more than 1-D arrays. Only takes in a single axis parameter (rather than `axes` list in other distance metrics).  There is likely a better way to handle `nil` options with `NimbleOptions` so suggestions on that are welcome.

### CODE REVIEW

File: lib/scholar/metrics/distance.ex
Comment (polvalente) [inline lib/scholar/metrics/distance.ex:433]:
```suggestion
  $hamming(x ,y) = \frac{#unequal(x, y)}{#x}$
```

File: lib/scholar/metrics/distance.ex
Comment (seanmor5) [inline lib/scholar/metrics/distance.ex:447]:
```suggestion
      >

```

File: lib/scholar/metrics/distance.ex
Comment (polvalente) [inline lib/scholar/metrics/distance.ex:479]: Why can't we support :axes instead of just a single axis?

File: lib/scholar/metrics/distance.ex
Comment (seanmor5) [inline lib/scholar/metrics/distance.ex:470]:
```suggestion
    if options[:weights] do
```

nil is falsey so I think you can flip this clause and remove

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:479]: This was mainly due to `weighted_mean` only taking `axis` as a param. This brings up additional questions as to whether we'd want to support multi-dim tensors (scipy elects to only support 1D). The current `moduledoc` for `Metrics.Distance` reads `Distance metrics between 1-D tensors.`

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:470]: Agreed, I will change that.

File: lib/scholar/metrics/distance.ex
Comment (polvalente) [reply lib/scholar/metrics/distance.ex:479]: Perhaps weighted_mean could be extended for having N-dimensional weights?

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:433]: The denominator is really the length of the axis you're summing over. If we wanted to allow no axis, `Nx.mean` would then use the total number of items so again the denominator would be $m \times n$. I'm not sure of a succinct way of expressing this

Comment (polvalente) [reply lib/scholar/metrics/distance.ex:433]: I've edited the suggestion to mention the aggregation axes.

I think for non-weighted we can support many axes and extend the weighted version when weighted_mean supports them.

Comment (polvalente) [reply lib/scholar/metrics/distance.ex:433]: Also, `#` here symbolizes the size of the set

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:479]: @seanmor5 I think this is more of a question for you

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:433]: Ah, thanks I didn't realize that about `#`. I'll add in support for many axes in non-weighted

File: lib/scholar/metrics/distance.ex
Comment (polvalente) [reply lib/scholar/metrics/distance.ex:479]: It was kind of a self-question as well haha!

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:433]: @polvalente Do you think it makes sense to change the `axes` validator to allow for `int`s for a single axis? Then within the validator just wrap the int in a list and pass that to the function for use as `:axes` parameter? This seems more intuitive than requiring a list of one item (but then again, the term `axes` suggests an enumerable type)

Comment (polvalente) [reply lib/scholar/metrics/distance.ex:433]: I'd rather always have lists, mainly because it's the way that the :axes parameter works in all of Nx so far

Comment (acalejos) [reply lib/scholar/metrics/distance.ex:433]: makes sense, thanks

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:479]: `Nx.mean` accepts `axes` so it seems like it would be consistent to have it for `weighted_mean` as well

Comment (polvalente) [reply lib/scholar/metrics/distance.ex:480]: @acalejos https://github.com/elixir-nx/nx/pull/1043 adds support for `:axes` instead of `:axis` in `Nx.weighted_mean`

File: lib/scholar/metrics/distance.ex
Comment (polvalente) [inline lib/scholar/metrics/distance.ex:485]: I just realized we can't have `weights` as an option here.
This is because options are enclosed in most of our compilation infrastructure, so tensors must be passed as proper arguments to be kept as tensor expressions.

The code to deal with this is unfortunately a bit unwieldly, but it's basically like this:

```elixir
defmodule T do
import Nx.Defn

deftransform hamming(x, y), do: non_weighted(x, y)
deftransform hamming(x, y, opts) when is_list(opts) do
  # call nimble opts here :)
  non_weighted(x, y, opts)
end

deftransform hamming(x, y, w), do: weighted(x, y, w)
deftransform hamming(x, y, w, opts) when is_list(opts) do
  # call nimble opts here :)
  weighted(x, y, w, opts)
end

defnp weighted(x, y, w, opts \\ []) do
   ...
end

defnp non_weighted(x, y, opts \\ []) do
...
end
end
```

the guard on the second clause is what enables us to have both hamming/2 (non-weighted) with optional opts and hamming/3 (weighted) with optional opts

File: lib/scholar/options.ex
Comment (polvalente) [inline lib/scholar/options.ex:14]: given my latest review, we don't need this function anymore

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:480]: Oh nice. Should this wait on that getting merged into Nx then?

Comment (polvalente) [reply lib/scholar/metrics/distance.ex:480]: You can point to my branch while it's not merged so you can keep working on it :)

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:485]: I think i see what youre saying. I think weights make more sense as a tensor so that's probably the way to go rather than having them be a list and stay in options.

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex:480]: What do you mean exactly by point to your branch? Would I do that from the mix.exs file?

Comment (polvalente) [reply lib/scholar/metrics/distance.ex:480]: Yes!

You can change the nx declaration in mix.exs to:

`{:nx, ">= 0.0.0", github: "elixir-nx/nx", sparse: "nx", branch: "pv-feat/multi-axes-weighted-mean"}`

This is basically saying: fetch from the elixir-nx/nx repo, only the nx directory, and checkout the given branch

File: lib/scholar/metrics/distance.ex
Comment (acalejos) [reply lib/scholar/metrics/distance.ex]: Made the changes at 7f4c606fdd861c60237af92d416d48e34de21016

Using the new `weighted_mean` with `axes` parameter some other Scholar tests failed. Seems to be only the PCA tests

Comment (polvalente) [reply lib/scholar/metrics/distance.ex]: The failed tests are actually due to SVD. I'll update Scholar's main branch with Nx main

Comment (polvalente) [reply lib/scholar/metrics/distance.ex]: Done :)
You can update your fork with Scholar main and it'll pull latest Nx as a dependency

Comment (acalejos) [reply lib/scholar/metrics/distance.ex]: Looks like I had to run mix format. I pushed again after pulling from Scholar/main

File: lib/scholar/metrics/distance.ex
Comment (polvalente) [inline lib/scholar/metrics/distance.ex:431]:
```suggestion
      iex> weights = Nx.tensor([1, 0.5, 0.5])
```

File: lib/scholar/metrics/distance.ex
Comment (polvalente) [inline lib/scholar/metrics/distance.ex:451]:
```suggestion
  deftransform hamming(x, y), do: hamming_unweighted(x, y)
```

File: lib/scholar/metrics/distance.ex
Comment (polvalente) [inline lib/scholar/metrics/distance.ex:458]:
```suggestion
  deftransform hamming(x, y, w), do: hamming_weighted(x, y, w)
```

**Review by polvalente — APPROVED** [review]

**Comment by polvalente** [comment]
Don't worry about the formatter! I'll fix this in a few minutes :)

**Comment by acalejos** [comment]
> ut
It failed a check:
```
warning: this clause for __defn:hamming__/2 cannot match because a previous clause at line 451 always matches
Warning:   lib/scholar/metrics/distance.ex:451
```

I thought that the guard would make these function signatures distinct?

**Comment by polvalente** [comment]
> It failed a check:
> 
> ```
> warning: this clause for __defn:hamming__/2 cannot match because a previous clause at line 451 always matches
> Warning:   lib/scholar/metrics/distance.ex:451
> ```
> 
> I thought that the guard would make these function signatures distinct?

This is probably a bug on how `deftransform` is defined. I'll try to fix this as well upstream.

**Comment by acalejos** [comment]
> This is probably a bug on how deftransform is defined. I'll try to fix this as well upstream.

I'll be interested to see how you fix it. Surprised that hasn't come up before

**Comment by polvalente** [comment]
@acalejos this seems to fix it https://github.com/elixir-nx/nx/pull/1045

Let's hold off until the fix is merged

File: lib/scholar/metrics/distance.ex
Comment (josevalim) [inline lib/scholar/metrics/distance.ex:465]: Declaring `opts` as default argument is no longer necessary as of latest Nx.

```suggestion
  defnp hamming_unweighted(x, y, opts) do
```

File: lib/scholar/metrics/distance.ex
Comment (josevalim) [inline lib/scholar/metrics/distance.ex:470]:
```suggestion
  defnp hamming_weighted(x, y, w, opts) do
```

File: lib/scholar/metrics/distance.ex
Comment (josevalim) [inline lib/scholar/metrics/distance.ex:451]: This should also work, no?

```suggestion
  deftransform hamming(x, y, opts \\ [])
```

File: lib/scholar/metrics/distance.ex
Comment (josevalim) [inline lib/scholar/metrics/distance.ex:460]:
```suggestion
  deftransform hamming(x, y, w, opts \\ []) when is_list(opts) do
```

Comment (polvalente) [reply lib/scholar/metrics/distance.ex:460]: Will this work though? Because this third argument can be either a list or a tensor depending on the clause

Comment (polvalente) [reply lib/scholar/metrics/distance.ex:460]: Nevermind, I misread this as being the arity-3 version

File: lib/scholar/metrics/distance.ex
Comment (polvalente) [reply lib/scholar/metrics/distance.ex:451]: The problem here is that the 3rd argument can either be a list or a tensor

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**
